crates/biome_js_analyze/src/analyzers/suspicious/no_confusing_void_type.rs

```diff
@@ -53,88 +53,100 @@ declare_rule! {
         let node = ctx.query();
-        return Some(
-            RuleDiagnostic::new(
-                rule_category!(),
-                node.range(),
-                markup! {{match_message(state)}},
-            )
-            .note(markup! {
-                "Remove "<Emphasis>"void"</Emphasis>
-            }),
-        );
+        let message = match state {
+            VoidTypeContext::Union => "void is not valid as a constituent in a union type",
+            VoidTypeContext::Unknown => {
+                "void is only valid as a return type or a type argument in generic type"
+            }
+        };
+
+        Some(
+            RuleDiagnostic::new(rule_category!(), node.range(), markup! {{message}}).note(
+                markup! {
+                    "Remove "<Emphasis>"void"</Emphasis>
+                },
+            ),
+        )
     }
 }
 
-fn node_in(node: &SyntaxNode<Language>) -> Option<VoidTypeIn> {
+fn decide_void_type_context(node: &SyntaxNode<JsLanguage>) -> Option<VoidTypeContext> {
     for parent in node.parent()?.ancestors() {
         match parent.kind() {
+            JsSyntaxKind::TS_UNION_TYPE_VARIANT_LIST => {
+                // checks if the union type contains a generic type has a void type as argument
+                for child in parent.descendants() {
+                    if child.kind() == JsSyntaxKind::TS_TYPE_ARGUMENT_LIST {
+                        let found_void_type = child
+                            .descendants()
+                            .any(|descendant| descendant.kind() == JsSyntaxKind::TS_VOID_TYPE);
+                        if found_void_type {
+                            return None;
+                        }
+                    }
+                }
+            }
+
             // (string | void)
-            // string | void
             // string & void
             // arg: void
             // fn<T = void>() {}
             JsSyntaxKind::TS_PARENTHESIZED_TYPE
-            | JsSyntaxKind::TS_UNION_TYPE_VARIANT_LIST
             | JsSyntaxKind::TS_INTERSECTION_TYPE_ELEMENT_LIST
             | JsSyntaxKind::TS_TYPE_ANNOTATION
             | JsSyntaxKind::TS_DEFAULT_TYPE_CLAUSE => {
                 continue;
             }
 
             JsSyntaxKind::TS_UNION_TYPE => {
-                return Some(VoidTypeIn::Union);
+                return Some(VoidTypeContext::Union);
+            }
+
+            // Promise<void>
+            JsSyntaxKind::TS_TYPE_ARGUMENT_LIST => {
+                // handle generic function that has a void type argument
+                // e.g. functionGeneric<void>(undefined);
+                let Some(grand_parent) = parent.grand_parent() else {
+                    continue;
+                };
+                if grand_parent.kind() == JsSyntaxKind::JS_CALL_EXPRESSION {
+                    return Some(VoidTypeContext::Unknown);
+                }
+                return None;
```

Comment:
We need to address the scenario where `fn<void>(undefined)`.
I discovered this issue when I adde tests, so I addressed.

Reply:
What the issue?

Reply:
I think this `void` in function's generic type may confuse developers because the `void` is useless to be generic.